google/cloud/forseti/scanner/audit/retention_rules_engine.py

```diff
@@ -0,0 +1,318 @@
+        """
+        super(RetentionRuleBook, self).__init__()
+        self._rules_sema = threading.BoundedSemaphore(value=1)
+
+        self.resource_rules_map = {
+            applies_to: collections.defaultdict(set)
+            for applies_to in _APPLY_TO_RESOURCES}
+        if not rule_defs:
+            self.rule_defs = {}
+        else:
+            self.rule_defs = rule_defs
+            self.add_rules(rule_defs)
+
+    def add_rules(self, rule_defs):
+        """Add rules to the rule book
+
+        Args:
+            rule_defs (dict): rule definitions dictionary
+        """
+        for (i, rule) in enumerate(rule_defs.get('rules', [])):
+            self.add_rule(rule, i)
+
+    def add_rule(self, rule_def, rule_index):
+        """Add a rule to the rule book.
+
+        Args:
+            rule_def (dict): A dictionary containing rule definition
+                properties.
+            rule_index (int): The index of the rule from the rule definitions.
+                Assigned automatically when the rule book is built.
+        """
+        self._rules_sema.acquire()
+        try:
+            applies_to = rule_def.get('applies_to', None)
+            if applies_to is None:
+                raise audit_errors.InvalidRulesSchemaError(
+                    'Lack of applies_to in rule {}'.format(rule_index))
+            if type(applies_to).__name__ != 'list':
+                raise audit_errors.InvalidRulesSchemaError(
+                    'Miss dash (-) near applies_to in rule {}'.format(
+                        rule_index))
+
+            minimum_retention = rule_def.get('minimum_retention', None)
+            maximum_retention = rule_def.get('maximum_retention', None)
+            if minimum_retention is None and maximum_retention is None:
+                raise audit_errors.InvalidRulesSchemaError(
+                    'Lack of minimum_retention and \
+maximum_retention in rule {}'.format(rule_index))
+            elif minimum_retention != None and maximum_retention != None:
+                if minimum_retention > maximum_retention:
+                    raise audit_errors.InvalidRulesSchemaError(
+                        'minimum_retention larger than \
+maximum_retention in rule {}'.format(rule_index))
+
+            applies_to_history = {}
+            for appto in applies_to:
+                if appto in applies_to_history:
+                    raise audit_errors.InvalidRulesSchemaError(
+                        'Duplicate applies_to in rule {}'.format(rule_index))
+                applies_to_history[appto] = True
+
+                if appto == _APPLY_TO_BUCKETS:
+                    self.add_bucket_rule(
+                        rule_def,
+                        rule_index,
+                        minimum_retention,
+                        maximum_retention)
+                # other appto add here
+        finally:
+            self._rules_sema.release()
+
+    def add_bucket_rule(self, rule_def, rule_index,
```

Comment:
The rules for each applies_to type will probably look the same, so you can make this function general (with applies_to as a param) and then have

```
self.resource_rules_map[applies_to][gcp_resource].add(rule)
```